feat: add Algolia DocSearch to documentation site

giladgd opened the pull request:
# Description
### Added Algolia DocSearch to the documentation site
> The screenshots below use the testing `apiKey` provided by Algolia

In order for the search to actually work you have to do the following:
1. Go to https://docsearch.algolia.com/apply/ and register TypeORM, it's free for open-source projects.
2. Edit `apiKey`, `appId` and `indexName` in `script/component/MainPage.js` with the values provided to you by Algolia.

Closes: #16 

### Changed Vue router's mode from "hash" to "history" to support search engine indexing
Previously site URLs looked like this:
```txt
https://typeorm.io/#/data-source
https://typeorm.io/#/data-source/how-to-use-datasource
```

Now site URLs will look like this:
```txt
https://typeorm.io/data-source
https://typeorm.io/data-source#how-to-use-datasource
```

The previous URLs will continue to work but will now redirect to the new URLs.
In order to prevent getting a 404 error when opening a deep URL, a custom `404.html` page is implemented to redirect back to `index.html` which will then use the history API to load the original URL without a refresh.

Based on #36

# Screenshots
![localhost_3000_ (2)](https://user-images.githubusercontent.com/7817232/159367936-2130403f-1b22-467e-a89c-64524b3b027f.png)

![localhost_3000_ (3)](https://user-images.githubusercontent.com/7817232/159367951-c667d33e-809b-4bc8-92c8-497b1c00f66b.png)

![localhost_3000_ (4)](https://user-images.githubusercontent.com/7817232/159367968-5341703a-fe5a-486c-a482-7644ed6f7a19.png)

giladgd committed:
feat: add Algolia DocSearch to documentation site

The `apiKey`, `appId` and `indexName` parameters specified in `MainPage.js` should be updated in order for the search to work

giladgd commented:
@pleerock Please let me know if I could help in any way to merge this feature

pleerock commented:
Here is what I got from the email:

CSS
Copy this snippet at the end of the HTML <head> tag
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@docsearch/css@3"/></pre></li>

JavaScript
Copy this snippet at the end of the HTML <body> tag
<script src="https://cdn.jsdelivr.net/npm/@docsearch/js@3"></script>

<script type="text/javascript">

  docsearch({

    appId: 2UUKVSBT3M,

    apiKey: 8392de2c2e76d873b60bf10562aab788,

    indexName: typeorm,

    container: '### REPLACE ME WITH A CONTAINER (e.g. div) ###'

    debug: false // Set debug to true if you want to inspect the modal

  });

</script>

I guess you can now finish this PR.

giladgd committed:
fix: update Algolia DocSearch configuration

giladgd commented:
@pleerock I've updated the PR with the `apiKey`, but it seems that the data index should also be configured at Algolia so the search can find results also from pages other than just the main page.
You can experiment with its configuration [here](https://www.algolia.com/users/sign_in) and [here](https://crawler.algolia.com/).

If you like, you can add me to the TypeORM index team at Algolia ([here](https://www.algolia.com/account/teams)) so I can experiment with the configuration and make it work well with the way the documentation site works

giladgd committed:
feat: change Vue router's mode from "hash" to "history" to support search engine indexing

In order to support loading a deep url without getting a 404 error, a custom 404.html page is implemented to redirect back to index.html which uses history api to load the original url without a refresh.
Based on #36

giladgd commented:
@pleerock I think setting [this config](https://www.algolia.com/doc/tools/crawler/apis/configuration/render-java-script/) would make the search work across all pages after merging this PR

giladgd committed:
fix: main page title links

pleerock commented:
thank you!

giladgd commented:
@pleerock It seems like the search doesn't index the documentation correctly.
Can you please add me to the documentation crawler account so I can try to fix it?
You can do it here: https://crawler.algolia.com/

giladgd commented:
My email is giladgd(at)gmail(dot)com

pleerock commented:
@giladgd done!

giladgd commented:
@pleerock The search is working now, you can remove me from the crawler account now if you like.
PR #57 should add proper titles to some results.
